wiki/docs/modules/authentication.md

```diff
@@ -0,0 +1,199 @@
+```
+  async generateToken(@Req() req: Request, @Res() res: Response): Promise<Response> {
+    try {
+      const user: ValidateUserDto = req.body;
+      const resWithAcessToken = await this.authService.generateToken(user);
+      return res.success(resWithAcessToken, StatusCodes.OK);
+    } catch (e) {
+      return res.error(e);
+    }
+  }
+```
+- The `AuthService` class looks like:
+
+```
+async generateToken(user: ValidateUserDto): Promise<any> {
+    try {
+      const userData = await this.findUserByEmail(user);
+      const isMatched = await comparePassword(user.password, userData.password);
+      if(isMatched) {
+        const payload = `${userData.firstName}${userData.id}`;
+        const accessToken = this.jwtService.sign(payload);
+        return {
+          access_token: accessToken,
+        };
+      } else {
+        throw new HttpException('Invalid credentials', HttpStatus.UNAUTHORIZED);
+      }
+    } catch (e) {
+      throw e;
+    }
+  }
+```
+- The `generateToken` method of the `AuthService` class created the JWT after the user credentials are validated. If the credentials are not validated then error will be thrown.
+- The `sign` method from `@nestjs/jwt` generates the token.
+- The token which is generated will be in the below format:
+
+- Image
+
+## STEP 2: CONFIGURATION INSIDE AUTH MODULE.
+
+- The configuration inside the `auth.module.ts` is as follows.
+```
+@Module({
+  imports: [
+    UsersModule,
+    JwtModule.registerAsync({
+      imports: [ConfigModule],
+      useFactory: async (config: ConfigService) => {
+        return {
+          secret: config.get<string>('SECRET_JWT_KEY'),
+        };
+      },
+      inject: [ConfigService],
+    }),
+    ConfigModule,
+    PassportModule,
+  ],
+  providers: [AuthService, JwtStrategy],
+  controllers: [AuthController],
+  // exports:[AuthService]
+})
+export class AuthModule {}
+``` 
+- Here, we are basically registering the `JwtModule`. While doing so, we specify the `secret` key.
+- The value of the secret key is been specified in the `.env` file.
+
+## STEP 3: IMPLEMENTING THE JWT PASSPORT STRATEGY
+
+- We need to do this to be able to address our second requirement of protecting endpoints using JWT.
+- We will create a new file `jwt.strategy.ts` within the auth folder and place the below code.
+
+```
+import { Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { PassportStrategy } from '@nestjs/passport';
+import { ExtractJwt, Strategy } from 'passport-jwt';
+
+@Injectable()
+export class JwtStrategy extends PassportStrategy(Strategy) {
+  constructor(private config: ConfigService) {
+    super({
+      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+      ignoreExpiration: false,
+      secretOrKey: config.get('auth.secretKey')
+    });
+  }
+
+  async validate(payload: any) {
+    return {
+      firstName: payload.firstName,
+      id: payload.id,
+    };
+  }
+}
```

Comment:
What is the meaning of returning firstName and id?

Reply:
We are not using this object anywhere but the Passport library is using it for validation from the access token.